modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/SqlQueryProcessor.java

```diff
@@ -317,6 +333,66 @@ public CompletableFuture<ExecutionTarget> forSystemView(ExecutionTargetFactory f
         services.forEach(LifecycleAware::start);
     }
 
+    // need to be refactored after TODO: https://issues.apache.org/jira/browse/IGNITE-20925
+    /** Get primary replicas. */
+    private CompletableFuture<List<NodeWithTerm>> primaryReplicas(int tableId) {
+        int catalogVersion = catalogManager.latestCatalogVersion();
+
+        Catalog catalog = catalogManager.catalog(catalogVersion);
+
+        CatalogTableDescriptor tblDesc = Objects.requireNonNull(catalog.table(tableId), "table");
+
+        CatalogZoneDescriptor zoneDesc = Objects.requireNonNull(catalog.zone(tblDesc.zoneId()), "zone");
+
+        int partitions = zoneDesc.partitions();
+
+        List<CompletableFuture<NodeWithTerm>> result = new ArrayList<>(partitions);
+
+        HybridTimestamp clockNow = clock.now();
+
+        // no need to wait all partitions after pruning was implemented.
+        for (int partId = 0; partId < partitions; ++partId) {
+            int partitionId = partId;
+            ReplicationGroupId partGroupId = new TablePartitionId(tableId, partitionId);
+
+            CompletableFuture<ReplicaMeta> f = placementDriver.awaitPrimaryReplica(
+                    partGroupId,
+                    clockNow,
+                    AWAIT_PRIMARY_REPLICA_TIMEOUT,
+                    SECONDS
+            );
+
+            result.add(f.handle((primaryReplica, e) -> {
+                if (e != null) {
+                    LOG.error("Failed to retrieve primary replica for partition {}", e, partitionId);
+
+                    throw withCause(IgniteInternalException::new, REPLICA_UNAVAILABLE_ERR, "Failed to get the primary replica"
+                            + " [tablePartitionId=" + partGroupId + ']', e);
+                } else {
+                    String holder = primaryReplica.getLeaseholder();
+
+                    assert holder != null : "Unable to map query, nothing holds the lease";
+
+                    ClusterNode node = clusterSrvc.topologyService().getByConsistentId(holder);
+
+                    if (node == null) {
+                        // additional recovery logic is need to be present around here.
+                        throw new IgniteInternalException(Sql.MAPPING_ERR, "Unable to map query, node is lost or offline");
+                    }
```

Comment:
do we really need to resolve node?

Reply:
yep, seems we can remove this part, done